File: iOS/Kalorie/Core/UseCases/AddFavouriteFoodUseCase.swift
//
//  AddFavouriteFoodUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 12.08.2026.
//

import Foundation

protocol AddFavouriteFoodUseCaseProtocol {
    func callAsFunction(_ item: FoodItemDomain) async throws
}

struct AddFavouriteFoodUseCase: AddFavouriteFoodUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(_ item: FoodItemDomain) async throws {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let dto = FavouriteFoodDTO(item: item, favouritedAt: .now)
        try await dataProvider.setAsync(dto, id: item.id, in: Constants.Firestore.favouriteFoods(userId: userId))
    }
}

#if DEBUG
struct AddFavouriteFoodUseCaseFake: AddFavouriteFoodUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(_ item: FoodItemDomain) async throws {
        if shouldThrow { throw URLError(.unknown) }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/AddFavouriteFoodUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.networking.FavouriteFoodDTO
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants
import java.time.Instant

interface AddFavouriteFoodUseCaseProtocol {
    suspend operator fun invoke(item: FoodItemDomain)
}

class AddFavouriteFoodUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : AddFavouriteFoodUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(item: FoodItemDomain) {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val dto = FavouriteFoodDTO(item = item, favouritedAt = Instant.now())
        dataProvider.setAsync(dto, id = item.id, inCollection = Constants.Firestore.favouriteFoods(userId))
    }
}
